stdlib/array.gr

```diff
@@ -262,3 +268,57 @@ export let count = (fn, array) => {
   }
   count
 }
+
+export let counti = (fn, array) => {
+  let length = length(array)
+  let mut position = 0
+  let mut count = 0
+  while (position < length) {
+    if(fn(array[position], position)) {
+      count = incr(count)
+      position = incr(position)
+    } else {
+      position = incr(position)
+    }
+  }
+  count
+}
+
+export let filter = (fn, array) => {
```

Comment:
I know this seems sort of annoying, but we can't really call `fn` more than once because of the possible side effects. It's really hard to write `Array.filter` efficiently, and that's probably why other functional languages don't really have it. What you could do instead is first make a copy of the array, place the filtered items at the start of the copied array, and then `init` a new array from the beginning of the copied array. It still kinda sucks because you need space for a full copy of the array, plus additional space for the array, potentially 3x the space of your array.

There's maybe the question of if you want to give your users a convenience function for something they probably shouldn't do much, as well. I don't have much of an opinion on this right now, but I'll chat with the core team and get back to you.

Reply:
We chatted, and we'll move forward with it!